spikeinterface/preprocessing/silence_periods.py

```diff
@@ -0,0 +1,132 @@
+                   basis
+    num_chunks_per_segment: int
+        Number of chunks per segment for random chunk, by default 20
+    chunk_size : int
+        Size of a chunk in number for random chunk, by default 10000
+    seed : int
+        Random seed for random chunk, by default None
+
+    Returns
+    -------
+    silence_recording: SilencedPeriodsRecording
+        The recording extractor after silencing some periods    
+    """
+    name = 'silence_periods'
+
+    def __init__(self, recording, list_periods, mode='zeros', num_chunks_per_segment=20,
+                 chunk_size=10000, seed=None):
+
+        available_modes = ('zeros', 'noise')
+        num_seg = recording.get_num_segments()
+
+
+        if num_seg == 1:
+            if isinstance(list_periods, (list, np.ndarray)) and not np.isscalar(list_periods[0]):
+                # when unique segment accept list instead of of list of list/arrays
+                list_periods = [list_periods]
+
+        # some checks
+        assert isinstance(list_periods, list), "'list_periods' must be a list (one per segment)"
+        assert len(list_periods) == num_seg, "'list_periods' must have the same length as the number of segments"
+        assert all(isinstance(list_periods[i], (list, np.ndarray)) for i in range(num_seg)), \
+            "Each element of 'list_periods' must be array-like"
+
+        for periods in list_periods:
+            if len(periods) > 0:
+                assert np.all(np.diff(np.array(periods), axis=1) > 0), "t_stops should be larger than t_starts"
+                assert all(e < s for (_, e), (s, _) in zip(periods, periods[1:])), "Intervals should not overlap"
+
+        assert mode in available_modes, f"mode {mode} is not an available mode: {available_modes}"
+
+        if mode in ['noise']:
+            noise_levels = get_noise_levels(recording, return_scaled=False, 
+                                    num_chunks_per_segment=num_chunks_per_segment,
+                                    chunk_size=chunk_size, concatenated=True, seed=seed)
+        else:
+            noise_levels = None
+
+        BasePreprocessor.__init__(self, recording)
+        for seg_index, parent_segment in enumerate(recording._recording_segments):
+            periods = list_periods[seg_index]
+            rec_segment = SilencedPeriodsRecordingSegment(parent_segment, periods, mode, noise_levels)
+            self.add_recording_segment(rec_segment)
+        
+        self._kwargs = dict(recording=recording.to_dict(), list_periods=list_periods,
+                            mode=mode, noise_levels=noise_levels)
+
+
+class SilencedPeriodsRecordingSegment(BasePreprocessorSegment):
+
+    def __init__(self, parent_recording_segment, periods, mode, noise_levels):
+
+        BasePreprocessorSegment.__init__(self, parent_recording_segment)
+
+        self.periods = np.asarray(periods, dtype='int64')
+        self.periods = np.sort(self.periods, axis=0)
```

Comment:
sort have a high cost when mutiprocessing! This should be externally or check at init.

Reply:
Done